code/go/internal/validator/semantic/validate_hbs_templates.go

```diff
@@ -0,0 +1,114 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package semantic
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"path"
+
+	"github.com/mailgun/raymond/v2"
+
+	"github.com/elastic/package-spec/v3/code/go/internal/fspath"
+	"github.com/elastic/package-spec/v3/code/go/internal/linkedfiles"
+	"github.com/elastic/package-spec/v3/code/go/pkg/specerrors"
+)
+
+var (
+	errInvalidHandlebarsTemplate = errors.New("invalid handlebars template")
+)
+
+// ValidateHandlebarsFiles validates all Handlebars (.hbs) files in the package filesystem.
+// It returns a list of validation errors if any Handlebars files are invalid.
+// hbs are located in both the package root and data stream directories under the agent folder.
+func ValidateHandlebarsFiles(fsys fspath.FS) specerrors.ValidationErrors {
+	// template files are placed at /agent/input directory or
+	// at the datastream /agent/stream directory
+	inputDir := path.Join("agent", "input")
+	err := validateTemplateDir(fsys, inputDir)
+	if err != nil {
+		return specerrors.ValidationErrors{
+			specerrors.NewStructuredErrorf("%w in %s: %w", errInvalidHandlebarsTemplate, inputDir, err),
+		}
+	}
+
+	datastreamEntries, err := fs.ReadDir(fsys, "data_stream")
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return specerrors.ValidationErrors{
+			specerrors.NewStructuredErrorf("error reading data_stream directory: %w", err),
+		}
+	}
+	for _, dsEntry := range datastreamEntries {
+		if !dsEntry.IsDir() {
+			continue
+		}
+		streamDir := path.Join("data_stream", dsEntry.Name(), "agent", "stream")
+		err := validateTemplateDir(fsys, streamDir)
+		if err != nil {
+			return specerrors.ValidationErrors{
+				specerrors.NewStructuredErrorf("%w in %s: %w", errInvalidHandlebarsTemplate, streamDir, err),
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateTemplateDir validates all Handlebars files in the given directory.
+func validateTemplateDir(fsys fspath.FS, dir string) error {
+	entries, err := fs.ReadDir(fsys, dir)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return err
+	}
+	for _, entry := range entries {
+		if path.Ext(entry.Name()) == ".hbs" {
+			err := validateHandlebarsEntry(fsys, dir, entry.Name())
+			if err != nil {
+				return err
+			}
+			continue
+		}
+		if path.Ext(entry.Name()) == ".link" {
+			linkFilePath := path.Join(dir, entry.Name())
+			linkFile, err := linkedfiles.NewLinkedFile(fsys.Path(linkFilePath))
+			if err != nil {
+				return err
+			}
+			err = validateHandlebarsEntry(fsys, dir, linkFile.IncludedFilePath)
+			if err != nil {
+				return err
```

Comment:
Same as above, it would be interesting to add the file name in the error:

```suggestion
				return fmt.Errorf("invalid template %q (%q): %w", entry.Name(), linkFile.IncludedFilePath, err)
```